File: rust/sdk/src/ledger_client.rs
use futures_core::Stream;
use m10_protos::sdk::{
    self, m10_query_service_client::M10QueryServiceClient,
    m10_tx_service_client::M10TxServiceClient, transaction_data::Data, BulkTransactions,
    TransactionData,
};
use m10_protos::sdk::{AccountInfo, RequestEnvelope};
use m10_signing::SignedRequest;
use std::time::{SystemTime, UNIX_EPOCH};
use tonic::{Request, Response};

use crate::transfer_ext::EnhancedTransfer;
use crate::{EnhancedTransferStep, Signer};
pub use tonic::transport::{Channel, ClientTlsConfig, Endpoint, Uri};

// Re-export public error
pub use tonic::Status;

#[derive(Clone)]
/// A client for the M10 Ledger.
///
/// This client allows you to query and transact on the M10 ledger.
///
/// # Example
/// ```no_run
/// #[tokio::main(flavor = "current_thread")]
/// async fn main() {
///   let ledger_url = "https://test.m10.net".to_string();
///   let mut client = m10_sdk::LedgerClient::new(
///     tonic::transport::Endpoint::from_shared(ledger_url)
///       .unwrap()
///       .connect_lazy()
///       .unwrap()
///    );
///
///   let block_height = client.block_height().await;
/// }
/// ```
pub struct LedgerClient {
    tx_client: M10TxServiceClient<Channel>,
    query_client: M10QueryServiceClient<Channel>,
}

impl LedgerClient {
    pub fn new(grpc_channel: Channel) -> Self {
        let tx_client = M10TxServiceClient::new(grpc_channel.clone());
        let query_client = M10QueryServiceClient::new(grpc_channel);
        Self {
            tx_client,
            query_client,
        }
    }

    pub fn transaction_request(
        data: impl Into<Data>,
        context_id: Vec<u8>,
    ) -> sdk::TransactionRequestPayload {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_micros() as u64;
        sdk::TransactionRequestPayload {
            nonce: fastrand::u64(..),
            timestamp,
            context_id,
            data: Some(TransactionData {
                data: Some(data.into()),
            }),
        }
    }

    pub async fn block_height(&mut self) -> Result<u64, tonic::Status> {
        let chain_info = self
            .query_client
            .get_chain_info(Request::new(()))
            .await?
            .into_inner();
        Ok(chain_info.block_height)
    }

    pub async fn create_transaction(
        &mut self,
        payload: impl Into<RequestEnvelope>,
    ) -> Result<sdk::TransactionResponse, Status> {
        self.tx_client
            .create_transaction(Request::new(payload.into()))
            .await
            .map(|res| res.into_inner())
    }

    pub async fn bulk_create_transactions(
        &mut self,
        payload: impl Into<BulkTransactions>,
    ) -> Result<sdk::BulkTransactionsResponse, Status> {
        self.tx_client
            .bulk_create_transactions(Request::new(payload.into()))
            .await
            .map(|res| res.into_inner())
    }

    // Transfers
    pub async fn get_transfer(
        &mut self,
        request: SignedRequest<sdk::GetTransferRequest>,
    ) -> Result<sdk::FinalizedTransfer, Status> {
        self.query_client
            .get_transfer(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    pub async fn list_transfers(
        &mut self,
        request: SignedRequest<sdk::ListTransferRequest>,
    ) -> Result<sdk::FinalizedTransfers, Status> {
        self.query_client
            .list_transfers(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    // Actions
    pub async fn get_action(
        &mut self,
        request: SignedRequest<sdk::GetActionRequest>,
    ) -> Result<sdk::Action, Status> {
        self.query_client
            .get_action(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    pub async fn list_actions(
        &mut self,
        request: SignedRequest<sdk::ListActionsRequest>,
    ) -> Result<sdk::Actions, Status> {
        self.query_client
            .list_actions(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    pub async fn get_transaction(
        &mut self,
        request: SignedRequest<sdk::GetTransactionRequest>,
    ) -> Result<sdk::FinalizedTransaction, Status> {
        self.query_client
            .get_transaction(Request::new(request.into()))
            .await
            .map(Response::into_inner)
    }

    pub async fn list_transactions(
        &mut self,
        request: SignedRequest<sdk::ListTransactionsRequest>,
    ) -> Result<sdk::FinalizedTransactions, Status> {
        self.query_client
            .list_transactions(Request::new(request.into()))
            .await
            .map(Response::into_inner)
    }

    pub async fn group_transactions(
        &mut self,
        request: SignedRequest<sdk::GroupTransactionsRequest>,
    ) -> Result<sdk::GroupedFinalizedTransactions, Status> {
        self.query_client
            .group_transactions(Request::new(request.into()))
            .await
            .map(Response::into_inner)
    }

    //  Indexed Accounts
    pub async fn get_indexed_account(
        &mut self,
        request: SignedRequest<sdk::GetAccountRequest>,
    ) -> Result<sdk::IndexedAccount, Status> {
        self.query_client
            .get_indexed_account(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    // AccountSets
    pub async fn get_account_set(
        &mut self,
        request: SignedRequest<sdk::GetAccountSetRequest>,
    ) -> Result<sdk::AccountSet, Status> {
        self.query_client
            .get_account_set(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    pub async fn list_account_sets(
        &mut self,
        request: SignedRequest<sdk::ListAccountSetsRequest>,
    ) -> Result<sdk::ListAccountSetsResponse, Status> {
        self.query_client
            .list_account_sets(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    // Accounts
    pub async fn get_account_metadata(
        &mut self,
        request: SignedRequest<sdk::GetAccountRequest>,
    ) -> Result<sdk::AccountMetadata, Status> {
        self.query_client
            .get_account_metadata(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    pub async fn get_account_info(
        &mut self,
        request: SignedRequest<sdk::GetAccountRequest>,
    ) -> Result<sdk::AccountInfo, Status> {
        self.query_client
            .get_account_info(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    pub async fn list_account_metadata(
        &mut self,
        request: SignedRequest<sdk::ListAccountMetadataRequest>,
    ) -> Result<sdk::ListAccountMetadataResponse, Status> {
        self.query_client
            .list_account_metadata(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    // Role Bindings
    pub async fn get_role_binding(
        &mut self,
        request: SignedRequest<sdk::GetRoleBindingRequest>,
    ) -> Result<sdk::RoleBinding, Status> {
        self.query_client
            .get_role_binding(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    pub async fn list_role_bindings(
        &mut self,
        request: SignedRequest<sdk::ListRoleBindingsRequest>,
    ) -> Result<sdk::ListRoleBindingsResponse, Status> {
        self.query_client
            .list_role_bindings(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    // Roles
    pub async fn get_role(
        &mut self,
        request: SignedRequest<sdk::GetRoleRequest>,
    ) -> Result<sdk::Role, Status> {
        self.query_client
            .get_role(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    pub async fn list_roles(
        &mut self,
        request: SignedRequest<sdk::ListRolesRequest>,
    ) -> Result<sdk::ListRolesResponse, Status> {
        self.query_client
            .list_roles(Request::new(request.into()))
            .await
            .map(|res| res.into_inner())
    }

    // Observations
    pub async fn observe_transfers(
        &self,
        request: SignedRequest<sdk::ObserveAccountsRequest>,
    ) -> Result<impl Stream<Item = Result<sdk::FinalizedTransactions, Status>>, Status> {
        self.query_client
            .clone()
            .observe_transfers(Request::new(request.into()))
            .await
            .map(tonic::Response::into_inner)
    }

    pub async fn observe_resources(
        &self,
        request: SignedRequest<sdk::ObserveResourcesRequest>,
    ) -> Result<impl Stream<Item = Result<sdk::FinalizedTransactions, Status>>, Status> {
        self.query_client
            .clone()
            .observe_resources(Request::new(request.into()))
            .await
            .map(tonic::Response::into_inner)
    }

    pub async fn observe_accounts(
        &self,
        request: SignedRequest<sdk::ObserveAccountsRequest>,
    ) -> Result<impl Stream<Item = Result<sdk::FinalizedTransactions, Status>>, Status> {
        self.query_client
            .clone()
            .observe_accounts(Request::new(request.into()))
            .await
            .map(tonic::Response::into_inner)
    }

    pub async fn observe_actions(
        &self,
        request: SignedRequest<sdk::ObserveActionsRequest>,
    ) -> Result<impl Stream<Item = Result<sdk::FinalizedTransactions, Status>>, Status> {
        self.query_client
            .clone()
            .observe_actions(Request::new(request.into()))
            .await
            .map(tonic::Response::into_inner)
    }

    pub async fn observe_metrics(
        &self,
        request: SignedRequest<sdk::ObserveAccountsRequest>,
    ) -> Result<impl Stream<Item = Result<sdk::TransactionMetrics, Status>>, Status> {
        self.query_client
            .clone()
            .observe_metrics(Request::new(request.into()))
            .await
            .map(tonic::Response::into_inner)
    }

    pub async fn enhance_transfers(
        &self,
        transfers: Vec<sdk::FinalizedTransfer>,
        signer: &impl Signer,
    ) -> Result<Vec<EnhancedTransfer>, Status> {
        futures_util::future::try_join_all(
            transfers
                .into_iter()
                .map(|transfer| self.enhance_transfer(transfer, signer)),
        )
        .await
    }

    pub async fn enhance_transfer(
        &self,
        transfer: sdk::FinalizedTransfer,
        signer: &impl Signer,
    ) -> Result<EnhancedTransfer, Status> {
        let mut enhanced_steps = Vec::default();
        for transfer_step in &transfer.transfer_steps {
            enhanced_steps.push(self.enhance_transfer_step(transfer_step, signer).await?);
        }
        Ok(EnhancedTransfer {
            enhanced_steps,
            transfer,
        })
    }

    async fn enhance_transfer_step(
        &self,
        transfer_step: &sdk::TransferStep,
        signer: &impl Signer,
    ) -> Result<EnhancedTransferStep, Status> {
        let from = async {
            Result::<_, Status>::Ok(
                self.query_client
                    .clone()
                    .get_account_info(Request::new(
                        signer
                            .sign_request(sdk::GetAccountRequest {
                                id: transfer_step.from_account_id.clone(),
                            })
                            .await
                            .map_err(|err| Status::internal(err.to_string()))?
                            .into(),
                    ))
                    .await
                    .map(|res| res.into_inner())
                    .ok(),
            )
        };
        let to = async {
            Result::<_, Status>::Ok(
                self.query_client
                    .clone()
                    .get_account_info(Request::new(
                        signer
                            .sign_request(sdk::GetAccountRequest {
                                id: transfer_step.to_account_id.clone(),
                            })
                            .await
                            .map_err(|err| Status::internal(err.to_string()))?
                            .into(),
                    ))
                    .await
                    .map(|res| res.into_inner())
                    .ok(),
            )
        };
        let (from, to) = futures_util::future::try_join(from, to).await?;
        let from_bank = async {
            if let Some(ref from) = from {
                if from.parent_account_id.is_empty() {
                    Ok(None)
                } else {
                    Result::<Option<AccountInfo>, Status>::Ok(
                        self.query_client
                            .clone()
                            .get_account_info(Request::new(
                                signer
                                    .sign_request(sdk::GetAccountRequest {
                                        id: from.parent_account_id.clone(),
                                    })
                                    .await
                                    .map_err(|err| Status::internal(err.to_string()))?
                                    .into(),
                            ))
                            .await
                            .map(|res| res.into_inner())
                            .ok(),
                    )
                }
            } else {
                Ok(None)
            }
        };
        let to_bank = async {
            if let Some(ref to) = to {
                if to.parent_account_id.is_empty() {
                    Ok(None)
                } else {
                    Result::<Option<AccountInfo>, Status>::Ok(
                        self.query_client
                            .clone()
                            .get_account_info(Request::new(
                                signer
                                    .sign_request(sdk::GetAccountRequest {
                                        id: to.parent_account_id.clone(),
                                    })
                                    .await
                                    .map_err(|err| Status::internal(err.to_string()))?
                                    .into(),
                            ))
                            .await
                            .map(|res| res.into_inner())
                            .ok(),
                    )
                }
            } else {
                Ok(None)
            }
        };
        let (from_bank, to_bank) = futures_util::future::try_join(from_bank, to_bank).await?;

        Ok(EnhancedTransferStep {
            from,
            to,
            from_bank,
            to_bank,
        })
    }

    pub async fn list_banks(
        &self,
        request: SignedRequest<sdk::ListBanksRequest>,
    ) -> Result<sdk::ListBanksResponse, Status> {
        self.query_client
            .clone()
            .list_banks(Request::new(request.into()))
            .await
            .map(tonic::Response::into_inner)
    }

    pub async fn get_bank(
        &self,
        request: SignedRequest<sdk::GetBankRequest>,
    ) -> Result<sdk::Bank, Status> {
        self.query_client
            .clone()
            .get_bank(Request::new(request.into()))
            .await
            .map(tonic::Response::into_inner)
    }
}

File: node/m10-sdk/src/ledger_client.ts
import * as grpc from "@grpc/grpc-js";

import { google, m10 } from "../protobufs";

import { M10Error } from "./error";
import type { EnhancedTransfer, EnhancedTransferStep } from "./transfer_ext";
import type { CryptoSigner } from "./utils";
import { getRPCImpl, getRPCImplStream, isSome, TransactionSigner } from "./utils";


/**
 * A client for the M10 Ledger.
 *
 * This client allows you to query and transact on the M10 ledger.
 *
 * @example
 * const ledgerUrl = "test.m10.net";
 * const ledgerClient = new LedgerClient(ledgerUrl, true);
 * const block_height = await ledgerClient.blockHeight();
 */
export class LedgerClient {

    private txClient: m10.sdk.M10TxService;
    private queryClient: m10.sdk.M10QueryService;

    private ledgerUrl: string;
    private credentials: grpc.ChannelCredentials;

    public constructor(ledgerUrl: string, tls: boolean) {
        this.ledgerUrl = ledgerUrl;
        this.credentials = tls ? grpc.credentials.createSsl() : grpc.credentials.createInsecure();

        this.txClient = m10.sdk.M10TxService.create(
            getRPCImpl(this.ledgerUrl, this.credentials, "m10.sdk.M10TxService"),
        );
        this.queryClient = m10.sdk.M10QueryService.create(
            getRPCImpl(this.ledgerUrl, this.credentials, "m10.sdk.M10QueryService"),
        );
    }

    private getQueryServiceStream(): m10.sdk.M10QueryService {
        return m10.sdk.M10QueryService.create(
            getRPCImplStream(this.ledgerUrl, this.credentials, "m10.sdk.M10QueryService"),
        );
    }

    // -------------------------------------------------------------------------
    // Transactions
    // -------------------------------------------------------------------------

    public transactionRequest(
        data: m10.sdk.transaction.ITransactionData,
        contextId?: Uint8Array,
    ): m10.sdk.transaction.ITransactionRequestPayload {
        return new m10.sdk.transaction.TransactionRequestPayload({
            data: data,
            nonce: TransactionSigner.generateNonce(),
            timestamp: TransactionSigner.getTimestamp(),
            contextId: contextId,
        });
    }

    public async blockHeight(): Promise<LongNumber> {
        const emptyRequest = new google.protobuf.Empty();
        const chainInfo = await this.queryClient.getChainInfo(emptyRequest);
        return chainInfo.blockHeight;
    }

    /**
     * @throws  {M10Error.Transaction} if response contains an error
     */
    public async createTransaction(
        signer: CryptoSigner,
        request: m10.sdk.transaction.ITransactionRequestPayload,
    ): Promise<m10.sdk.transaction.ITransactionResponse> {
        const payload = m10.sdk.transaction.TransactionRequestPayload.encode(request).finish();
        const envelop = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });
        const transactionResponse = await this.txClient.createTransaction(envelop);

        if (isSome(transactionResponse.error)) {
            throw M10Error.Transaction(transactionResponse.error);
        }

        return transactionResponse;
    }

    // -------------------------------------------------------------------------
    // Transfers
    // -------------------------------------------------------------------------

    public async getTransfer(
        signer: CryptoSigner,
        properties: m10.sdk.transaction.IGetTransferRequest,
    ): Promise<m10.sdk.transaction.IFinalizedTransfer> {

        const request = new m10.sdk.transaction.GetTransferRequest(properties);
        const payload = m10.sdk.transaction.GetTransferRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.getTransfer(envelope);
    }

    public async listTransfers(
        signer: CryptoSigner,
        properties: m10.sdk.transaction.IListTransferRequest,
    ): Promise<m10.sdk.transaction.IFinalizedTransfers> {

        const request = new m10.sdk.transaction.ListTransferRequest(properties);
        const payload = m10.sdk.transaction.ListTransferRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.listTransfers(envelope);
    }

    // -------------------------------------------------------------------------
    // Actions
    // -------------------------------------------------------------------------

    public async getAction(
        signer: CryptoSigner,
        properties: m10.sdk.transaction.IGetActionRequest,
    ): Promise<m10.sdk.transaction.IAction> {

        const request = new m10.sdk.transaction.GetActionRequest(properties);
        const payload = m10.sdk.transaction.GetActionRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.getAction(envelope);
    }

    public async listActions(
        signer: CryptoSigner,
        properties: m10.sdk.transaction.IListActionsRequest,
    ): Promise<m10.sdk.transaction.IActions> {

        const request = new m10.sdk.transaction.ListActionsRequest(properties);
        const payload = m10.sdk.transaction.ListActionsRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.listActions(envelope);
    }

    public async getTransaction(
        signer: CryptoSigner,
        properties: m10.sdk.IGetTransactionRequest,
    ): Promise<m10.sdk.IFinalizedTransaction> {

        const request = new m10.sdk.GetTransactionRequest(properties);
        const payload = m10.sdk.GetTransactionRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.getTransaction(envelope);
    }

    public async listTransactions(
        signer: CryptoSigner,
        properties: m10.sdk.IListTransactionsRequest,
    ): Promise<m10.sdk.IFinalizedTransactions> {

        const request = new m10.sdk.ListTransactionsRequest(properties);
        const payload = m10.sdk.ListTransactionsRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.listTransactions(envelope);
    }

    public async groupTransactions(
        signer: CryptoSigner,
        properties: m10.sdk.IGroupTransactionsRequest,
    ): Promise<m10.sdk.IGroupedFinalizedTransactions> {

        const request = new m10.sdk.GroupTransactionsRequest(properties);
        const payload = m10.sdk.GroupTransactionsRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.groupTransactions(envelope);
    }

    // -------------------------------------------------------------------------
    // Indexed Accounts
    // -------------------------------------------------------------------------

    public async getIndexedAccount(
        signer: CryptoSigner,
        properties: m10.sdk.transaction.IGetAccountRequest,
    ): Promise<m10.sdk.transaction.IIndexedAccount> {

        const request = new m10.sdk.transaction.GetAccountRequest(properties);
        const payload = m10.sdk.transaction.GetAccountRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.getIndexedAccount(envelope);
    }

    // -------------------------------------------------------------------------
    // AccountSets
    // -------------------------------------------------------------------------

    public async getAccountSet(
        signer: CryptoSigner,
        properties: m10.sdk.IGetAccountSetRequest,
    ): Promise<m10.sdk.model.IAccountSet> {

        const request = new m10.sdk.GetAccountSetRequest(properties);
        const payload = m10.sdk.GetAccountSetRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.getAccountSet(envelope);
    }


    public async listAccountSets(
        signer: CryptoSigner,
        properties: m10.sdk.IListAccountSetsRequest,
    ): Promise<m10.sdk.IListAccountSetsResponse> {

        const request = new m10.sdk.ListAccountSetsRequest(properties);
        const payload = m10.sdk.ListAccountSetsRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.listAccountSets(envelope);
    }

    // -------------------------------------------------------------------------
    // Accounts
    // -------------------------------------------------------------------------

    public async getAccountMetadata(
        signer: CryptoSigner,
        properties: m10.sdk.transaction.IGetAccountRequest,
    ): Promise<m10.sdk.model.IAccountMetadata> {

        const request = new m10.sdk.transaction.GetAccountRequest(properties);
        const payload = m10.sdk.transaction.GetAccountRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.getAccountMetadata(envelope);
    }

    public async getAccountInfo(
        signer: CryptoSigner,
        request: m10.sdk.transaction.GetAccountRequest,
    ): Promise<m10.sdk.model.IAccountInfo> {
        const payload = m10.sdk.transaction.GetAccountRequest.encode(request).finish();
        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });
        return this.queryClient.getAccountInfo(envelope);
    }

    public async listAccountMetadata(
        signer: CryptoSigner,
        request: m10.sdk.ListAccountMetadataRequest,
    ): Promise<m10.sdk.IListAccountMetadataResponse> {
        const payload = m10.sdk.ListAccountMetadataRequest.encode(request).finish();
        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });
        return this.queryClient.listAccountMetadata(envelope);
    }

    // -------------------------------------------------------------------------
    // Role Bindings
    // -------------------------------------------------------------------------

    public async getRoleBinding(
        signer: CryptoSigner,
        properties: m10.sdk.IGetRoleBindingRequest,
    ): Promise<m10.sdk.IRoleBinding> {

        const request = new m10.sdk.GetRoleBindingRequest(properties);
        const payload = m10.sdk.GetRoleBindingRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.getRoleBinding(envelope);
    }

    public async listRoleBindings(
        signer: CryptoSigner,
        properties: m10.sdk.IListRoleBindingsRequest,
    ): Promise<m10.sdk.IListRoleBindingsResponse> {

        const request = new m10.sdk.ListRoleBindingsRequest(properties);
        const payload = m10.sdk.ListRoleBindingsRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.listRoleBindings(envelope);
    }

    // -------------------------------------------------------------------------
    // Banks
    // -------------------------------------------------------------------------

    public async getBank(
        signer: CryptoSigner,
        request: m10.sdk.GetBankRequest,
    ): Promise<m10.sdk.model.IBank> {
        const payload = m10.sdk.GetBankRequest.encode(request).finish();
        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });
        return this.queryClient.getBank(envelope);
    }

    public async listBanks(
        signer: CryptoSigner,
        properties: m10.sdk.IListBanksRequest,
    ): Promise<m10.sdk.IListBanksResponse> {

        const request = new m10.sdk.ListBanksRequest(properties);
        const payload = m10.sdk.ListBanksRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        return this.queryClient.listBanks(envelope);
    }

    // -------------------------------------------------------------------------
    // Roles
    // -------------------------------------------------------------------------

    public async getRole(
        signer: CryptoSigner,
        request: m10.sdk.GetRoleRequest,
    ): Promise<m10.sdk.Role> {
        const payload = m10.sdk.GetRoleRequest.encode(request).finish();
        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });
        return this.queryClient.getRole(envelope);
    }

    public async listRoles(
        signer: CryptoSigner,
        request: m10.sdk.ListRolesRequest,
    ): Promise<m10.sdk.IListRolesResponse> {
        const payload = m10.sdk.ListRolesRequest.encode(request).finish();
        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });
        return this.queryClient.listRoles(envelope);
    }

    // -------------------------------------------------------------------------
    // Observations (can be improved after https://github.com/protobufjs/protobuf.js/pull/1115 is merged)
    // -------------------------------------------------------------------------

    public observeTransfers(
        signer: CryptoSigner,
        properties: m10.sdk.IObserveAccountsRequest,
    ): [m10.sdk.M10QueryService, () => void] {

        const request = new m10.sdk.ObserveAccountsRequest(properties);
        const payload = m10.sdk.ObserveAccountsRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        const service = this.getQueryServiceStream();
        const startObserver = (): void => { service.observeTransfers(envelope); };

        return [service, startObserver];
    }

    public observeResources(
        signer: CryptoSigner,
        properties: m10.sdk.IObserveResourcesRequest,
    ): [m10.sdk.M10QueryService, () => void] {

        const request = new m10.sdk.ObserveResourcesRequest(properties);
        const payload = m10.sdk.ObserveResourcesRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        const service = this.getQueryServiceStream();
        const startObserver = (): void => { service.observeResources(envelope); };

        return [service, startObserver];
    }

    public observeAccounts(
        signer: CryptoSigner,
        properties: m10.sdk.IObserveAccountsRequest,
    ): [m10.sdk.M10QueryService, () => void] {

        const request = new m10.sdk.ObserveAccountsRequest(properties);
        const payload = m10.sdk.ObserveAccountsRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        const service = this.getQueryServiceStream();
        const startObserver = (): void => { service.observeAccounts(envelope); };

        return [service, startObserver];
    }

    public observeActions(
        signer: CryptoSigner,
        properties: m10.sdk.IObserveActionsRequest,
    ): [m10.sdk.M10QueryService, () => void] {

        const request = new m10.sdk.ObserveActionsRequest(properties);
        const payload = m10.sdk.ObserveActionsRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        const service = this.getQueryServiceStream();
        const startObserver = (): void => { service.observeActions(envelope); };

        return [service, startObserver];
    }

    public observeMetrics(
        signer: CryptoSigner,
        properties: m10.sdk.IObserveAccountsRequest,
    ): [m10.sdk.M10QueryService, () => void] {

        const request = new m10.sdk.ObserveAccountsRequest(properties);
        const payload = m10.sdk.ObserveAccountsRequest.encode(request).finish();

        const envelope = new m10.sdk.RequestEnvelope({ payload, signature: signer.getSignature(payload) });

        const service = this.getQueryServiceStream();
        const startObserver = (): void => { service.observeMetrics(envelope); };

        return [service, startObserver];
    }

    // -------------------------------------------------------------------------
    // EnhancedTransfers
    // -------------------------------------------------------------------------

    public async enhanceTransfers(
        signer: CryptoSigner,
        transfers: m10.sdk.transaction.IFinalizedTransfer[],
    ): Promise<EnhancedTransfer[]> {

        const enhancedTransfersPromise = transfers.map((transfer) => this.enhanceTransfer(signer, transfer));
        const enhancedTransfers = await Promise.all(enhancedTransfersPromise);

        return enhancedTransfers;
    }

    public async enhanceTransfer(
        signer: CryptoSigner,
        transfer: m10.sdk.transaction.IFinalizedTransfer,
    ): Promise<EnhancedTransfer> {

        const enhancedTransferStepsPromise = (transfer.transferSteps || [])
            .map((transferStep) => this.enhanceTransferStep(signer, transferStep));

        const enhancedTransferSteps = await Promise.all(enhancedTransferStepsPromise);

        return {
            transfer: new m10.sdk.transaction.FinalizedTransfer(transfer),
            enhanced_steps: enhancedTransferSteps,
        };
    }

    public async enhanceTransferStep(
        signer: CryptoSigner,
        transferStep: m10.sdk.transaction.ITransferStep,
    ): Promise<EnhancedTransferStep> {

        const fromPromise = this.getAccountInfo(
            signer,
            new m10.sdk.transaction.GetAccountRequest({ id: transferStep.fromAccountId }),
        );
        const toPromise = this.getAccountInfo(
            signer,
            new m10.sdk.transaction.GetAccountRequest({ id: transferStep.toAccountId }),
        );

        const [from, to] = await Promise.all([fromPromise, toPromise]);

        const fromBankPromise = (
            isSome(from.parentAccountId)
                ? await this.getAccountInfo(signer, new m10.sdk.transaction.GetAccountRequest({ id: from.parentAccountId }))
                : null
        );
        const toBankPromise = (
            isSome(to.parentAccountId)
                ? await this.getAccountInfo(signer, new m10.sdk.transaction.GetAccountRequest({ id: to.parentAccountId }))
                : null
        );

        const [fromBank, toBank] = await Promise.all([fromBankPromise, toBankPromise].filter(isSome));

        return {
            from,
            to,
            from_bank: fromBank,
            to_bank: toBank,
        };
    }
}
